sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -192,104 +188,58 @@ class HashJoinIterator(
     private val spillCallback: SpillCallback,
     private val streamTime: GpuMetric,
     private val joinTime: GpuMetric,
-    private val totalTime: GpuMetric) extends Iterator[ColumnarBatch] with Arm with Logging {
-  import scala.collection.JavaConverters._
-
-  // For some join types even if there is no stream data we might output something
-  private var initialJoin = true
-  // If the join explodes this holds batches from the stream side split into smaller
-  // pieces.
-  private val pendingSplits = mutable.Queue[SpillableColumnarBatch]()
-  private var nextCb: Option[ColumnarBatch] = None
-  private var gathererStore: Option[JoinGatherer] = None
-  // Close the input data, the lazy spillable batch now owns it.
-  private val built = withResource(builtInput) { builtInput =>
-    LazySpillableColumnarBatch(builtInput, spillCallback, "built")
-  }
-
+    private val totalTime: GpuMetric)
+    extends SplittableJoinIterator(
+      s"hash $joinType gather",
+      stream,
+      streamAttributes,
+      built,
+      targetSize,
+      spillCallback,
+      joinTime = joinTime,
+      streamTime = streamTime,
+      totalTime = totalTime) {
   // We can cache this because the build side is not changing
-  private lazy val estimatedRowsPerStreamBatch = joinType match {
+  private lazy val streamMagnificationFactor = joinType match {
     case _: InnerLike | LeftOuter | RightOuter =>
       withResource(GpuProjectExec.project(built.getBatch, boundBuiltKeys)) { builtKeys =>
-        guessJoinRowsForTargetSize(builtKeys)
+        guessStreamMagnificationFactor(builtKeys)
       }
     case _ =>
       // existence joins don't change size, and FullOuter cannot be split
       1.0
   }
-  private var closed = false
-
-  def close(): Unit = {
-    if (!closed) {
-      built.close()
-      nextCb.foreach(_.close())
-      nextCb = None
-      gathererStore.foreach(_.close())
-      gathererStore = None
-      pendingSplits.foreach(_.close())
-      pendingSplits.clear()
-      closed = true
-    }
-  }
-
-  TaskContext.get().addTaskCompletionListener[Unit](_ => close())
-
-  private def nextCbFromGatherer(): Option[ColumnarBatch] = {
-    withResource(new NvtxWithMetrics("hash join gather", NvtxColor.DARK_GREEN, joinTime)) { _ =>
-      val ret = gathererStore.map { gather =>
-        val nextRows = JoinGatherer.getRowsInNextBatch(gather, targetSize)
-        gather.gatherNext(nextRows)
-      }
-      if (gathererStore.exists(_.isDone)) {
-        gathererStore.foreach(_.close())
-        gathererStore = None
-      }
 
-      if (ret.isDefined) {
-        // We are about to return something. We got everything we need from it so now let it spill
-        // if there is more to be gathered later on.
-        gathererStore.foreach(_.allowSpilling())
-      }
-      ret
+  override def computeNumJoinRows(cb: ColumnarBatch): Long = {
+    // TODO: Replace this estimate with exact join row counts using the corresponding cudf APIs
```

Comment:
Is there a follow on issue for this, or should it be done now?

Reply:
The Java APIs do not exist yet, so it can't be done now.  I'll update this TODO to point to the cudf issue tracking the Java bindings for this.